File: nestjs/src/decorators.ts
import { ToolDefinition, ToolVersion } from './models';

export function tool(options?: {
  name?: string;
  description?: string;
  author?: string;
  version?: string;
  sourceRepo?: string;
  sourceRef?: string;
  dependencies?: string[];
}) {
  return function (
    target: unknown,
    propertyKey?: string,
    descriptor?: PropertyDescriptor,
  ) {
    const func = descriptor ? descriptor.value : target;
    if (typeof func !== 'function') {
      throw new Error('@tool decorator must be applied to a function or method');
    }

    const toolName = options?.name || func.name || 'anonymous';
    const toolDesc = options?.description || toolName;
    const version = options?.version || '1.0.0';

    const parameters = inferParameters(func);

    const toolVersion: ToolVersion = {
      version,
      source_repo: options?.sourceRepo || '',
      source_ref: options?.sourceRef || '',
      dependencies: options?.dependencies || [],
      released_at: new Date().toISOString(),
      is_prerelease: /alpha|beta|rc/.test(version.toLowerCase()),
    };

    const toolDef: ToolDefinition = {
      name: toolName,
      description: toolDesc,
      parameters,
      author: options?.author || 'anonymous',
      entry_point: `${func.name}`,
      versions: [toolVersion],
    };

    (func as Record<string, unknown>).__funchub_tool__ = toolDef;

    if (descriptor) {
      return descriptor;
    }
    return func;
  };
}

function inferParameters(func: Function): Record<string, unknown> {
  const str = func.toString();
  const paramsMatch = str.match(/\(([^)]*)\)/);
  if (!paramsMatch) {
    return { type: 'object', properties: {} };
  }

  const paramStr = paramsMatch[1];
  const properties: Record<string, Record<string, unknown>> = {};
  const required: string[] = [];

  const paramList = paramStr.split(',').map((p) => p.trim()).filter(Boolean);

  for (const param of paramList) {
    const parts = param.split('=');
    const name = parts[0].trim().replace(/^\.\.\./, '');
    if (name === 'this') continue;

    const prop: Record<string, unknown> = { type: 'string' };
    if (parts.length > 1) {
      prop.default = parts[1].trim();
    } else {
      required.push(name);
    }
    properties[name] = prop;
  }

  const schema: Record<string, unknown> = { type: 'object', properties };
  if (required.length > 0) {
    schema.required = required;
  }
  return schema;
}

File: python/funchub/decorators.py
import inspect
from typing import Any, Callable, Dict, List, Optional, Type, get_type_hints

from funchub.models import ToolDefinition, ToolVersion

_TYPE_MAP: Dict[type, str] = {
    str: "string",
    int: "integer",
    float: "number",
    bool: "boolean",
    list: "array",
    dict: "object",
    type(None): "null",
}


def _py_type_to_json_type(py_type: Type) -> str:
    return _TYPE_MAP.get(py_type, "string")


def _infer_parameters(func: Callable) -> Dict[str, Any]:
    sig = inspect.signature(func)
    hints = get_type_hints(func)
    properties: Dict[str, Dict[str, Any]] = {}
    required: List[str] = []

    for name, param in sig.parameters.items():
        if name in ("self", "cls"):
            continue
        param_type = hints.get(name, str)
        json_type = _py_type_to_json_type(param_type)
        prop: Dict[str, Any] = {"type": json_type}

        if param.default is not inspect.Parameter.empty:
            prop["default"] = param.default
            default_str = str(param.default)
            if json_type == "string":
                prop["description"] = f"Default: {default_str}"
        else:
            required.append(name)

        properties[name] = prop

    schema: Dict[str, Any] = {
        "type": "object",
        "properties": properties,
    }
    if required:
        schema["required"] = required
    return schema


def tool(
    name: Optional[str] = None,
    description: Optional[str] = None,
    author: str = "anonymous",
    version: str = "1.0.0",
    source_repo: str = "",
    source_ref: str = "",
    dependencies: Optional[List[str]] = None,
):
    def decorator(func: Callable) -> Callable:


        tool_name = name or func.__name__
        tool_desc = description or (func.__doc__ or "").strip() or tool_name
        parameters = _infer_parameters(func)
        tool_version = ToolVersion(
            version=version,
            source_repo=source_repo,
            source_ref=source_ref,
            dependencies=dependencies or [],
            is_prerelease="alpha" in version or "beta" in version or "rc" in version,
        )
        tool_def = ToolDefinition(
            name=tool_name,
            description=tool_desc,
            parameters=parameters,
            author=author,
            entry_point=f"{func.__module__}:{func.__name__}",
            versions=[tool_version],
        )
        func.__funchub_tool__ = tool_def
        return func

    return decorator


funchub_tool = tool
